### PR TITLE
Fix AOTI integer symbol recovery from composed sizes

### LINKED ISSUE #185806: [Inductor][AOTI] codegen_symbol uses float reciprocal for integer dim recovery, silently corrupts dynamic batches when 1/K is not exactly representable in IEEE-754
### 🐛 Describe the bug

`torch._inductor.aoti_compile_and_package` produces a package whose compiled forward silently returns wrong output rows whenever it has to recover a symbolic batch dim `s` from an input shape `K*s` and `(1.0/K)*K != 1.0` in IEEE-754 double precision. For non-power-of-2 `K` (e.g. `K=1496`, common in vision-tower token-grid shapes), the corruption hits at multiple runtime batch sizes — including the most degenerate case `bs=1`, which crashes inside `flash_attention` because the recovered batch evaluates to `0`.

Eager forward of the same `nn.Module` is correct at every batch on every version tested.

The bug is a composition of two PRs that are each individually correct:
- **#95698** (Mar 2023): made `CppPrinter._print_Rational` emit float division, intentionally, to fix `upsample_bilinear2d` / `upsample_nearest` dynamic-shape regressions (those ops need real fractional scale factors).
- **#146044** (Jan 2025): added composed-constraint dim recovery in `cpp_wrapper_cpu.py:codegen_symbol`, lowering `try_solve` results through the same `CppPrinter`. This routes integer dim recovery through the float-div printer.

Issue #145500 (the motivating bug for #146044) only used `K=2`, which is exactly representable, so the bug never surfaced in regression tests.

### Minimal reproducer (~30 lines, no extra deps)

```python
import os
import torch, torch.nn as nn, torch.nn.functional as F
from torch.export import Dim

D    = int(os.environ.get("D", 1280))
H    = int(os.environ.get("H", 16))
HD   = int(os.environ.get("HD", 80))
L    = int(os.environ.get("L", 1496))    # any K such that (1.0/K)*K != 1.0 reproduces
MAXN = int(os.environ.get("MAXN", 8))
assert D == H * HD


class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.qkv = nn.Linear(D, 3 * D)
        self.out = nn.Linear(D, D)
    def forward(self, x, sizes):
        torch._check(x.shape[0] == sizes.shape[0] * L)
        x = x.unflatten(0, (-1, L))
        q, k, v = self.qkv(x).reshape(-1, L, 3, H, HD).permute(2, 0, 3, 1, 4).unbind(0)
        x = F.scaled_dot_product_attention(q, k, v).transpose(1, 2).reshape(-1, L, D)
        return self.out(x.flatten(0, 1))


dev, dt = "cuda", torch.bfloat16
m = M().to(dev, dt).eval()
ex = (torch.randn(4 * L, D, device=dev, dtype=dt),
      torch.zeros(4, dtype=torch.long, device=dev))
ep = torch.export.export(m, ex, dynamic_shapes={"x": {0: Dim.AUTO}, "sizes": {0: Dim.AUTO}}, strict=True)
torch._inductor.aoti_compile_and_package(ep, package_path="/tmp/repro.pt2")
aoti = torch._inductor.aoti_load_package("/tmp/repro.pt2")

print(f"torch {torch.__version__}  example bs=4  L={L}  MAXN={MAXN}")
with torch.inference_mode():
    for bs in range(1, MAXN + 1):
        x = torch.randn(bs * L, D, device=dev, dtype=dt)
        s = torch.zeros(bs, dtype=torch.long, device=dev)
        eager_rows = m(x, s).shape[0]
        try:
            aoti_rows = aoti(x, s).shape[0]
            tag = "OK " if aoti_rows == eager_rows else "BAD"
        except Exception:
            aoti_rows, tag = "CRASH", "ERR"
        print(f"  bs={bs}: eager={eager_rows:4d}  aoti={aoti_rows!s:>5}  {tag}")
```

### Observed output (torch 2.9.1+cu128, RTX 3090)

```
torch 2.9.1+cu128  example bs=4  L=1496  MAXN=8
  bs=1: eager=1496  aoti=CRASH  ERR     ← flash_attn: "batch size must be positive" (recovered batch == 0)
  bs=2: eager=2992  aoti= 1496  BAD     ← one image short
  bs=3: eager=4488  aoti= 4488  OK
  bs=4: eager=5984  aoti= 4488  BAD     ← one image short
  bs=5: eager=7480  aoti= 7480  OK
  bs=6: eager=8976  aoti= 8976  OK
  bs=7: eager=10472  aoti=10472  OK
  bs=8: eager=11968  aoti=10472  BAD    ← one image short
```

Runtime warning that points at the root cause:

```
[W Resize.cpp:31] Warning: An output with one or more elements was resized since it had
shape [1496, 3840], which does not match the required output shape [2992, 3840].
```

`[L, 3*D]` = one image's worth of rows = AOTI's compile-time symbolic size for the QKV linear output. The runtime eager fallback resizes it because the true shape needs `bs*L` rows.

### Same bug reproduces on every version tested

| torch | bs=1 | bs=2 | bs=3 | bs=4 | bs=5 | bs=6 | bs=7 | bs=8 |
|---|---|---|---|---|---|---|---|---|
| 2.8.0+cu128 | CRASH | bad | ✓ | bad | ✓ | ✓ | ✓ | bad |
| 2.9.1+cu128 | CRASH | bad | ✓ | bad | ✓ | ✓ | ✓ | bad |
| 2.12.0.dev20260408+cu128 | bad (0 rows) | bad | ✓ | bad | ✓ | ✓ | ✓ | bad |

(2.12 nightly silently returns 0 rows at `bs=1` instead of crashing — same bug, slightly different surface.)

### Pure-Python predictor (no torch needed)

```python
def will_aoti_be_wrong_at(L: int, N: int) -> bool:
    return int((1.0 / L) * (N * L)) != N
```

Verified L-sensitivity (matches AOTI output exactly):

| L | rationale | predicted-bad bs∈[1..16] | observed-bad bs |
|---|---|---|---|
| 1024 | `1/L = 2^-10` exactly representable | none | none ✓ |
| 1000 | `(1/L)*L` round-trips to exactly 1.0 | none | none ✓ |
| 1496 | `(1/L)*L = 0.9999999999999999` | {1, 2, 4, 8, 16} | {1, 2, 4, 8, 16} ✓ |
| 1500 | `(1/L)*L` round-trips to exactly 1.0 | none | none ✓ |

(`L=1496` is `34*44`, the Qwen2-VL 640×480 grid — a real shape, not contrived.)

#### Bad C++ in `wrapper.cpp:331` of the bs=1 crash trace

```cpp
int64_t s46 = (1.0/1496.0)*arg4_1_size_0;
```

This recovers the symbolic batch `s46` from `arg4_1.size(0) == 1496*s46` by floating-point reciprocal multiplication and truncation to `int64_t`. `1.0/1496.0` is not exactly representable in IEEE-754, so the product accumulates rounding error and truncates wrong for many inputs:

```
bs=1: q*1496  = 0.9999999999999999       → int64_t cast = 0  → flash_attn crash
bs=2: q*2992  = 1.9999999999999998       → 1                  → 1 image short
bs=3: q*4488  = exactly 3.0              → 3                  → ✓
bs=4: q*5984  = 3.9999999999999996       → 3                  → 1 image short
...
bs=8: q*11968 = 7.999999999999999        → 7                  → 1 image short
```

The wrong `s46` then propagates through every downstream buffer allocation (`{1496L*s46, 3840L}` for the QKV output, etc.).

#### Where this gets emitted

**`torch/_inductor/codegen/cpp_wrapper_cpu.py:331`** (`codegen_symbol`) recovers free symbols from input shapes:

```python
solution = try_solve(sympy.Eq(sym_or_exp, size_symbol), free_symbol)
if solution is not None:
    code.writeline(f"int64_t {free_symbol} = {cexpr(solution[1])};")
```

`try_solve` returns the algebraically correct `Mul(Rational(1, 1496), arg4_1_size_0)`. The expression then goes through `cexpr` (the `CppPrinter`), whose `_print_Rational` at **`torch/utils/_sympy/printers.py:400`** is documented to emit float division:

```python
def _print_Rational(self, expr):
    # Uses float constants to perform FP div
    if expr.q == 1:
        r = f"{expr.p}"
    else:
        r = f"{expr.p}.0/{expr.q}.0"
    return f"static_cast<{INDEX_TYPE}>({r})" if expr.is_integer else r
```

Combined with `ExprPrinter._print_Mul` (just `stringify(args, "*")`), `Mul(Rational(1, 1496), x)` lowers to `(1.0/1496.0)*x`. The outer `int64_t {free_symbol} = …;` then truncates the float to integer, dropping the rounding error into the symbol.

#### Why this was deliberate (and why it's still wrong here)

PR #95698 explicitly switched `_print_Rational` from integer division to FP division to fix `upsample_bilinear2d` / `upsample_nearest` dynamic shapes — those ops need real fractional scale factors `output_h / input_h`. The change was correct for upsample.

PR #146044 reused the same `cexpr` printer for integer dim recovery in AOTI, without realizing that the printer's `Rational` handling assumes float context. Issue #145500 (the motivating bug) used `K=2`, which is exactly representable in IEEE-754 — so the test never tripped the rounding path.

### Proposed fix

**Patch the call site, not the printer.** The type information ("I'm writing into `int64_t`") only exists at `codegen_symbol`; the printer is shared with upsample and other float-context callers.

```python
# torch/_inductor/codegen/cpp_wrapper_cpu.py:codegen_symbol

solution = try_solve(sympy.Eq(sym_or_exp, size_symbol), free_symbol)
if solution is not None:
    expr = solution[1]
    # try_solve returns Mul(Rational(1, K), x); rewrite to CleanDiv so cexpr
    # emits c10::div_floor_integer (exact integer division) instead of
    # (1.0/K.0)*x (FP reciprocal that loses precision when 1/K is not
    # exactly representable in IEEE-754).
    if isinstance(expr, sympy.Mul):
        coeff, rest = expr.as_coeff_Mul()
        if isinstance(coeff, sympy.Rational) and coeff.p == 1 and coeff.q != 1:
            from torch.utils._sympy.functions import CleanDiv
            expr = CleanDiv(rest, coeff.q)
    code.writeline(f"int64_t {free_symbol} = {cexpr(expr)};")
    bound_vars.add(free_symbol)
```

`CleanDiv` is torch's "exact integer division, divisor evenly divides" SymPy op — the constraint `K*s == size` from the export graph guarantees the divisibility. Its existing `CppPrinter._print_FloorDiv` (`printers.py:424`) emits:

```cpp
c10::div_floor_integer(static_cast<int64_t>(arg4_1_size_0), static_cast<int64_t>(1496L))
```

Why patch here and not `try_solve` or `_print_Rational`:

- `try_solve` has 15+ callers across torch (export, ShapeEnv, dynamo, range analysis). Most use the result purely symbolically — they shouldn't be perturbed.
- `_print_Rational` is load-bearing for upsample's fractional scale factors. Touching it regresses PR #95698's original motivation.
- Only `codegen_symbol` (and one or two similar call sites in `tiling_utils.py`) knows the lowering target is `int64_t`. Type-aware lowering belongs at the call site.

Patch was applied locally and verified.

### Verification (post-patch, torch 2.9.1+cu128)

**Patched `wrapper.cpp` line emitted:**

```cpp
int64_t s46 = c10::div_floor_integer(static_cast<int64_t>(arg4_1_size_0), static_cast<int64_t>(1496L));
```

(was: `int64_t s46 = (1.0/1496.0)*arg4_1_size_0;`)

**Minimal repro, L=1496, MAXN=16:**

```
torch 2.9.1+cu128  example bs=4  L=1496  MAXN=16
  bs=1..16: eager == aoti for every batch.  (all OK)
```

**Real-world repro** (Qwen2-VL vision tower at 640×480, 32 attention layers, SDPA + rotary):

```
post-patch: bs=1..8 all OK.
```

**Regression checks:**

- `L=1024`: still all OK (was already correct; patch doesn't disturb it).
- `L=1000`: still all OK (lucky round-trip preserved).
- Upsample dynamic shapes: unchanged (the patch only touches the `Mul(Rational(1,K), x)` pattern coming from `codegen_symbol`'s solver path; upsample's float scale factors flow through different ops and are not affected).

### Versions

```
PyTorch version: 2.9.1+cu128
CUDA used to build PyTorch: 12.8
OS: Linux (x86_64) 6.8.0-84-generic
Python: 3.12.11
GPU: NVIDIA GeForce RTX 3090
cuDNN: 9.8.0

Versions of relevant libraries:
[pip3] torch==2.9.1+cu128
[pip3] triton==3.5.1
```

Same bug reproduces identically on:
- 2.8.0+cu128
- 2.12.0.dev20260408+cu128 (nightly)

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @iupaikov-amd

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185841

AOTI C++ wrapper input symbol binding can recover a free integer symbol by
solving a composed input size expression such as 1496*s == arg0_size_0.
try_solve returns the algebraically correct rational expression
arg0_size_0/1496, but codegen_symbol sent that expression directly through the
shared C++ SymPy printer. That printer intentionally emits rationals as floating
point division for float-context users like upsample scale factors, so the
wrapper generated `(1.0/1496.0) * arg0_size_0` and assigned the rounded double
to int64_t. For divisors whose reciprocal does not round-trip, valid batch sizes
such as 1, 2, 4, and 8 truncate to one less.

Keep the rational printer unchanged and make the integer lowering context
explicit at the call site. Before printing a solved integer symbol recovery,
normalize the solved expression with SymPy and rewrite any rational denominator
to CleanDiv. CleanDiv uses the existing exact integer division C++ lowering,
which preserves the shape constraint semantics expected by export.

The alternative of changing CppPrinter._print_Rational would regress existing
float-context users. Changing try_solve would also perturb symbolic callers that
are not lowering into an int64_t declaration.

Fixes #185806
Generated by my agent

Benchmark Results:
A tiny C++ microbenchmark comparing only the generated wrapper arithmetic showed
no negative signal for the new expression:

- old float reciprocal expression: 126696471 ns total, 0.633482 ns/iter
- new positive integer division expression: 54508727 ns total, 0.272544 ns/iter

The expression is evaluated once per recovered runtime symbol per wrapper
invocation, so end-to-end model impact should be negligible.

Test Plan:
- Reproduced the pre-patch issue with a small AOTI dynamic-shape module on CPU
  and CUDA; bs=1,2,4,8 recovered one fewer batch row.
- Reran the same repro after the fix on CPU and CUDA; bs=1,2,3,4,8 matched eager
  output.
- Ran an equivalent targeted AOTI check through test_aot_inductor_utils on CPU
  and CUDA; generated C++ contained c10::div_floor_integer, did not contain
  (1.0/1496.0), and runtime outputs matched eager for bs=1,2,3,4,8.
- Attempted `python test/inductor/test_aot_inductor.py -k composed_dynamic_size_integer_symbol_recovery -v`; blocked during import by the local incompatible torchvision install (`operator torchvision::nms does not exist`).
- Ran helper-level SymPy/C++ printer checks for arg0_1_size_0/1496,
  arg0_1_size_0/1496 - 1, (arg0_1_size_0 + 3)/1496, and arg0_1_size_0/748.
- Ran `git diff --check`.
- Ran `lintrunner -a`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main to pick up the docs workflow fix for the stale linux-docs upload failures (`aws: command not found`). The PR code did not need changes for those failures.

**Comment by desertfire** [comment]
@jansel  Let me know when this is ready for review.

**Comment by jansel** [comment]
@desertfire

**Review by desertfire — APPROVED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
My agent says the remaining failed Windows build jobs are runner setup failures (`pip` cannot use SSL before PyTorch builds), not failures from this PR. I asked mergebot to ignore them with `@pytorchbot merge -i`.
